### ISSUE TITLE
[Performance] ApplyTopKTopPCustom kernel: ScatterCumtomImpl per-element scatter writes cause 100-239ms latency when top_p < 1.0 and top_k has no effective truncation

### ISSUE THREAD

**Issue by xdchenming-gif** [vllm-project/vllm-ascend#10821]
## Summary

The `ApplyTopKTopPCustom` AscendC custom kernel used for top-k/top-p sampling in vllm-ascend exhibits a severe performance degradation: **~1000x latency increase** (from <1ms to 100-239ms per call) when `top_p < 1.0` is combined with `top_k` having no effective truncation (i.e., `top_k >= vocab_size`). In production workloads on Ascend 910B2C with Qwen3-32B (vocab=151936), this kernel dominates engine step time, consuming up to **73.3% of step latency** (256.4ms out of 349.8ms).

## Environment

| Item | Value |
|------|-------|
| Model | Qwen3-32B |
| Hardware | 8 × Ascend 910B2C / pod |
| Engine | vLLM V1 (vLLM `0.18.0` + vllm-ascend `0.18.0`) |
| CANN | 8.5.0 |
| Deployment | 200 pods, TP=8, PP=1, `--no-async-scheduling` |
| Python | 3.11.14 |

The issue also exists in vllm-ascend **v0.18.0** release, which uses the same `ApplyTopKTopPCustom` kernel implementation.

## Problem Description

In high-batch decode scenarios, the `ApplyTopKTopPCustom` operator intermittently shows ~1000x latency divergence:

| Path | ApplyTopKTopPCustom avg | Share of Computing | Trigger condition |
|------|------------------------|-------------------|-------------------|
| **Slow** | 30-239ms/call | 20-72% of device ops | `top_p < 1.0` + no effective `top_k` truncation |
| **Fast** | 0.03-0.25ms/call | <1% | Normal requests with effective `top_k` or no `top_p` |

The intermittency is not random hardware jitter: it is determined by whether the current decode batch contains requests with specific sampling parameter combinations (`top_p < 1.0` + `top_k >= vocab_size`). Even a single request with `top_p=0.9` in a batch of 384 can drag the entire batch into the slow path.

### Kernel Entry Dispatch

The custom op dispatches to three paths based on tiling key (`csrc/apply_top_k_top_p_custom/op_kernel/apply_top_k_top_p_custom.cpp:26-41`):

```cpp
if (TILING_KEY_IS(0)) {
    op.Process();       // top-k + top-p combined
} else if (TILING_KEY_IS(1)) {
    op.ProcessTopK();   // top-k only
} else if (TILING_KEY_IS(2)) {
    op.ProcessTopP();   // top-p only
}
```

The tiling decision in `op_host/apply_top_k_top_p_custom_tiling.cpp:149-150`:
- `onlyTopK_ = 1` when `k` is present but `p` is absent (tilingKey=1 → `ProcessTopK()`)
- `onlyTopP_ = 2` when `p` is present but `k` is absent (tilingKey=2 → `ProcessTopP()`)  
- When both `k` and `p` are present: tilingKey=0 → **`Process()` combined path**

This means when vLLM passes both `k` and `p` tensors (which happens whenever any request in the batch has `top_p < 1.0`), the kernel always enters `Process()`, even if `top_k >= vocab_size` makes the top-k constraint a no-op.

### The Slow Path: `ScatterCumtomImpl()`

The core bottleneck is in `ScatterCumtomImpl()` (`apply_top_k_top_p_custom.h:395-409`):

```cpp
void ScatterCumtomImpl(uint32_t loopBatch, uint32_t loopProbNum, uint32_t offset) {
    for (int32_t loopProb = 0; loopProb < loopProbNum; loopProb++) {
        float cumsumDataTmp = cumSumRes.GetValue(loopProb);
        if (cumsumDataTmp <= pValue) {
            continue;    // ★ continue, NOT break — scans entire chunk
        }
        scatterTensor.SetValue(0, sortedValueLocal[offset].GetValue(loopProb));  // Scalar
        int32_t gmIndex = sortedIndicesLocal[offset].GetValue(loopProb);         // Scalar: random target
        SToMTE3Sync();   // ★ Full pipeline stall: Scalar→MTE3
        DataCopyPad(mGmOut_[baseGmIdx_ + gmIndex],
                    scatterTensor.ReinterpretCast<outputT>(),
                    {1, sizeof(outputT), 0, 0, 0});   // 4-byte DMA to random HBM address
        MTE3ToSSync();   // ★ Full pipeline stall: MTE3→Scalar
    }
}
```

Three structural problems:

1. **Full chunk scan with `continue` instead of `break`**: The loop scans the entire chunk, only skipping low-probability tokens with `continue`. It cannot early-terminate.

2. **Per-element 4-byte random scatter write**: Each retained token triggers a `DataCopyPad` of just 4 bytes to a random vocab position (`mGmOut_[baseGmIdx_ + gmIndex]`).

3. **Dual sync barriers per element**: `SToMTE3Sync()` + `MTE3ToSSync()` serialize the Scalar and MTE3 pipelines for every single retained token.

### Contrast with the Fast Path: `ScatterCumtomImplTopK()`

The top-k-only path (`apply_top_k_top_p_custom.h:635-650`) avoids this problem:

```cpp
void ScatterCumtomImplTopK(uint32_t loopBatch, uint32_t loopProbNum, uint32_t offset) {
    for (int32_t loopProb = loopProbNum - 1; loopProb >= 0; loopProb--) {
        float curValue = calLocalFp32[offset].GetValue(loopProb);
        if (curValue < kthValue) {
            break;   // ★ Reverse traversal + break! Early termination
        }
        // scatter (same per-element, but only processes ~top_k tokens ≈ 40)
    }
}
```

With effective `top_k=50`, only ~40 scatter operations occur (~6us). With no effective `top_k` and `top_p=0.9`, retained tokens can reach tens of thousands to full-vocab, resulting in ~174ms.

### PMU Signature Evidence

| PMU Metric | Slow Path (iter7) | Fast Path (iter5) | Difference |
|------------|------------------|-------------------|------------|
| Duration | 175,774us | 189us | **930x** |
| MTE3 (UB→GM write) ratio | **63.6%** | 30.0% | 2.1x |
| Scalar ratio | **28.0%** | 33.1% | - |
| Vector ratio | 6.5% | **55.0%** | 0.12x |
| MTE3 time | 45,550us | 49us | **930x** |
| Scalar time | 20,034us | 54us | **371x** |

The slow path shows MTE3 (random writes) and Scalar (per-element index/condition) dominating, while Vector (actual computation) drops to near zero. This perfectly matches the `ScatterCumtomImpl()` code path.

**HBM bandwidth efficiency**: With 380 × 151936 × 4B = 220MB data and 910B2C's 392 GB/s HBM2e bandwidth, theoretical minimum time is ~0.56ms. Actual 175.8ms yields **~0.3% efficiency**.

### Direct-Op Microbenchmark

The slow path can be reproduced without a full vLLM service by calling the custom op directly on Ascend NPU:

```python
import torch
import torch_npu

# Use vllm-ascend's custom op
from vllm_ascend.utils import enable_custom_op
enable_custom_op()

batch, vocab = 384, 151936
logits = torch.empty((batch, vocab), dtype=torch.float32, device="npu:0")
logits.uniform_(-0.01, 0.01)  # Flat distribution amplifies top-p retained tokens

# Fast path: effective top_k truncation
k_50 = torch.full((batch,), 50, dtype=torch.int32, device="npu:0")
p_all = torch.full((batch,), 0.9, dtype=torch.float32, device="npu:0")
# top_k=50 + top_p=0.9 → ~4.7ms

# Slow path: top_k=vocab (no truncation) + top_p=0.9
k_vocab = torch.full((batch,), vocab, dtype=torch.int32, device="npu:0")
# top_k=151936 + top_p=0.9 → ~224ms
```

**Results** (batch=384, vocab=151936, warmup=1, iters=3):

| Case | top_k | top_p | Avg ms | Samples (ms) |
|------|-------|-------|--------|-------------|
| `k_only_50` | 50 | None | **4.667** | 4.669, 4.673, 4.658 |
| `p_all_k_vocab` | 151936 | 0.9 | **223.766** | 223.759, 223.741, 223.799 |
| `p_mixed_one_k_vocab` | 151936 | [0.9,1.0,...] | **246.140** | 246.139, 246.130, 246.151 |
| `p_all_k_50` | 50 | 0.9 | **4.729** | 4.726, 4.727, 4.733 |

Key observations:
- `top_k=50` keeps the kernel at ~4.7ms regardless of `top_p`
- `top_k=vocab` + `top_p=0.9` jumps to ~224ms
- Even 1 request with `top_p=0.9` in a batch of 384 can trigger the slow path for the entire batch

### Service-Level Reproduction

In a full vLLM service with `top_k=0, top_p=0.9, temperature=20.0`, profiler data confirms:

| Condition | Apply Shape | Apply Duration | PMU MTE3 Ratio |
|-----------|-------------|---------------|-----------------|
| `top_k=50` | 192×151936 | ~140us/call | No MTE3/Scalar amplification |
| `top_k=0`, B≈192 | 192×151936 | ~100,000us | MTE3 ratio ≈ 0.685 |
| `top_k=0`, B=384 | 384×151936 | ~200,000us | MTE3 ratio ≈ 0.685 |

### Fix A: Per-row fast path in kernel (Recommended, P0)

Add per-row logic to check if `top_p == 1.0 && top_k >= vocab_size` (no-op row) or `top_k < K_MAX` (effective top-k truncation). Currently, the tiling decision is whole-batch: a single request with `top_p < 1.0` forces the entire batch into the `Process()` combined path. The kernel should skip no-op rows and route effective-`top_k` rows to the top-k bounded scatter, even within `Process()`.

### Fix B: Tiling-level fast path (Recommended, P0)

In `SetTilingKey()`, when `top_k >= vocab_size` (i.e., top-k is a no-op), treat the combined case as top-p-only (tilingKey=2 → `ProcessTopP()`), since the top-k constraint has no effect. This avoids the `Process()` combined path entirely.

### Fix C: Vectorized mask-fill (P1, fundamental fix)

Instead of scatter-writing retained tokens back to random positions in the original vocab order, compute the top-p/top-k cutoff in the sorted domain and apply a vectorized mask-fill on the original logits:

```
sorted probs → compute cutoff → original order: logits < cutoff ? -inf : logits
```

This replaces O(retained_tokens) random 4B scatter writes with O(vocab/batch) contiguous vector operations, matching the behavior of the PyTorch fallback (`_apply_top_k_top_p_pytorch` in `sampler.py`).

### Fix D: Reduce per-element sync overhead (P1)

Batch retained (value, index) pairs in UB before performing scatter writes, reducing the `SToMTE3Sync/MTE3ToSSync` overhead from 2× per element to 2× per batch.

## Related PRs and Issues

| PR | Status | Description |
|-----|--------|-------------|
| #1920 | Merged (2025-07) | `npu_top_k_top_p` fused ops; requires `VLLM_ASCEND_ENABLE_TOPK_OPTIMIZE=1`, not enabled by default |
| #6098 | Merged (2026-01) | Removes k≤1024 limitation in tiling; does not fix the scatter slow path |
| #9948 | Open (2026-06-03) | Replace custom op with `torch_npu` native sampler; right direction, not yet merged |

## Workaround

Setting an effective `top_k` (e.g., `top_k=50` or `top_k=100`) for all requests prevents the slow path. This limits the retained token count and keeps `ApplyTopKTopPCustom` at ~4.7ms. However, this changes the candidate space and requires A/B quality evaluation.

Alternatively, enabling `VLLM_ASCEND_ENABLE_TOPK_OPTIMIZE=1` (PR #1920's fused ops) can bypass the custom op, but requires validation for precision and logprobs compatibility.

**Comment by Tflowers-0129** [comment]
@ZT-AIA Pls take a look about this issue.

**Comment by ZT-AIA** [comment]
This custom operator has been replaced by `torch_npu.npu_top_k_top_p`, which has the same functionality, better performance, and there are further optimization plans in the pipeline.https://github.com/vllm-project/vllm-ascend/pull/12232

**Comment by xdchenming-gif** [comment]
Thanks @ZT-AIA for the update. We noticed that PR #12232 replaces the custom AscendC operator with `torch_npu.npu_top_k_top_p`, which is a different approach from what we've been working on internally.

To provide some context: our team has also been optimizing the original `ApplyTopKTopPCustom` AscendC kernel. The key optimizations include:

1. **Non-discrete memory access**: Replace per-element scatter writes with `cmpsel`-based contiguous copies, avoiding the `SToMTE3Sync`/`MTE3ToSSync` pipeline stalls per element.
2. **Value-domain binary search** instead of sorting.
3. **K-value inversion** for large K scenarios.
4. **Special-case fast paths**: Direct copy when `top_p == 1.0` (60x improvement) or `top_k == vocab_size` (87x improvement).

Initial results show ~32x improvement for the `batch=128, P=1, K=V=129280` scenario. Full optimized code is planned for Aug 20, with complete benchmark testing by Sep 15, targeting the Q3 commercial release.

A few questions before we close this issue:

1. **Is PR #12232 already merged?** We'd like to verify `torch_npu.npu_top_k_top_p` performance on our production scale (Qwen3-32B, vocab=151936, batch up to 384, Ascend 910B2C with CANN 8.5.0).

2. **Precision compatibility**: Does `torch_npu.npu_top_k_top_p` produce identical results to the custom op for mixed `top_k`/`top_p` sampling batches? Our issue specifically highlights that even 1 request with `top_p=0.9` in a batch of 384 can trigger the slow path — we need to confirm the replacement handles mixed batches correctly.

3. **Relationship between the two tracks**: Are these two optimization paths complementary (the AscendC kernel optimization for cases where torch_npu can't cover), or is the intent to fully replace the custom op going forward?

Once we can validate PR #12232 against our production workload and confirm it resolves the 100-239ms latency, we're happy to close this issue.

**Comment by xdchenming-gif** [comment]
The R&D guy cxg told me that this issue can be closed, since CANN has reimplemented this operator.
btw ---  this behavior is not "community" at all.
